storage/replica.go

```diff
@@ -1755,7 +1755,12 @@ func isOnePhaseCommit(ba roachpb.BatchRequest) bool {
 	if _, hasBegin := ba.GetArg(roachpb.BeginTransaction); !hasBegin {
 		return false
 	}
-	if _, hasEnd := ba.GetArg(roachpb.EndTransaction); !hasEnd {
+	arg, hasEnd := ba.GetArg(roachpb.EndTransaction)
+	if !hasEnd {
+		return false
+	}
+	etArg := arg.(*roachpb.EndTransactionRequest)
+	if etArg.Deadline != nil && etArg.Deadline.Less(ba.Timestamp) {
```

Comment:
Agreed. Pushed another commit to refactor the preconditions. The commit defines  `isEndTransactionTriggeringRetryError` and `isEndTransactionExceedingDeadline` and uses them in `EndTransaction` and `isOnePhaseCommit`. It's a bit weird to have two methods, but more cleanup wasn't easy.. 

Could you take a look?